File: UNIT-01/orchestrator/llm_client.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net"
	"net/http"
	"time"
)

const (
	dialTimeout        = 10 * time.Second
	responseHeaderTimeout = 120 * time.Second
)

type LLMClient struct {
	endpoint   string
	model      string
	httpClient *http.Client
	showCache  *ModelShowResponse
}

type ModelShowResponse struct {
	Modelfile  string                 `json:"modelfile"`
	Parameters string                 `json:"parameters"`
	Template   string                 `json:"template"`
	Details    ModelDetails           `json:"details"`
	ModelInfo  map[string]interface{} `json:"model_info"`
}

type ModelDetails struct {
	ParentModel       string   `json:"parent_model"`
	Format            string   `json:"format"`
	Family            string   `json:"family"`
	Families          []string `json:"families"`
	ParameterSize     string   `json:"parameter_size"`
	QuantizationLevel string   `json:"quantization_level"`
}

func NewLLMClient(endpoint, defaultModel string) *LLMClient {
	if endpoint == "" {
		endpoint = "http://127.0.0.1:11434"
	}
	return &LLMClient{
		endpoint: endpoint,
		model:    defaultModel,
		httpClient: &http.Client{
			Transport: &http.Transport{
				DialContext: (&net.Dialer{
					Timeout: dialTimeout,
				}).DialContext,
				TLSHandshakeTimeout:   dialTimeout,
				ResponseHeaderTimeout: responseHeaderTimeout,
			},
		},
	}
}

// ShowModel fetches full model metadata from /api/show. Result is cached.
func (c *LLMClient) ShowModel() (*ModelShowResponse, error) {
	if c.showCache != nil {
		return c.showCache, nil
	}
	resp, err := c.httpClient.Post(c.endpoint+"/api/show", "application/json",
		bytes.NewBufferString(fmt.Sprintf(`{"name":"%s"}`, c.model)))
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	var show ModelShowResponse
	if err := json.NewDecoder(resp.Body).Decode(&show); err != nil {
		return nil, err
	}
	c.showCache = &show
	return c.showCache, nil
}

// directiveResult is the JSON structure returned by the model under schema constraint.
type directiveResult struct {
	Directives []struct {
		Name string                 `json:"name"`
		Args map[string]interface{} `json:"args"`
	} `json:"directives"`
}

// StreamDirectives sends a request with the directive JSON schema, returns parsed directives.
func (c *LLMClient) StreamDirectives(messages []ollamaMessage, temperature float64) ([]Directive, int, int, error) {
	reqBody := ollamaRequest{
		Model:    c.model,
		Messages: messages,
		Stream:   true,
		Format:   directiveSchema,
		Options: map[string]interface{}{
			"temperature": temperature,
		},
	}

	b, err := json.Marshal(reqBody)
	if err != nil {
		return nil, 0, 0, fmt.Errorf("marshal request: %w", err)
	}

	req, err := http.NewRequest("POST", c.endpoint+"/api/chat", bytes.NewReader(b))
	if err != nil {
		return nil, 0, 0, fmt.Errorf("create request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, 0, 0, fmt.Errorf("http: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		b, _ := io.ReadAll(resp.Body)
		return nil, 0, 0, fmt.Errorf("status %d: %s", resp.StatusCode, string(b))
	}

	fullText, _, promptTokens, outputTokens, err := ParseStreamCLI(resp.Body, io.Discard, nil)
	if err != nil {
		return nil, 0, 0, err
	}

	// Parse the JSON-structured output into directives.
	var result directiveResult
	if err := json.Unmarshal([]byte(fullText), &result); err != nil {
		return nil, promptTokens, outputTokens, fmt.Errorf("parse directive JSON: %w\nraw: %s", err, fullText)
	}

	directives := make([]Directive, 0, len(result.Directives))
	for _, d := range result.Directives {
		directives = append(directives, Directive{Name: d.Name, Args: d.Args})
	}
	return directives, promptTokens, outputTokens, nil
}

// StreamCLI sends a chat request to Ollama with streaming and directive parsing.
func (c *LLMClient) StreamCLI(messages []ollamaMessage, w io.Writer, stopSpinner func(), temperature float64) (string, []Directive, int, int, error) {
	reqBody := ollamaRequest{
		Model:    c.model,
		Messages: messages,
		Stream:   true,
		Options: map[string]interface{}{
			"temperature": temperature,
		},
	}

	b, err := json.Marshal(reqBody)
	if err != nil {
		return "", nil, 0, 0, fmt.Errorf("marshal request: %w", err)
	}

	req, err := http.NewRequest("POST", c.endpoint+"/api/chat", bytes.NewReader(b))
	if err != nil {
		return "", nil, 0, 0, fmt.Errorf("create request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return "", nil, 0, 0, fmt.Errorf("http: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		b, _ := io.ReadAll(resp.Body)
		return "", nil, 0, 0, fmt.Errorf("status %d: %s", resp.StatusCode, string(b))
	}

	return ParseStreamCLI(resp.Body, w, stopSpinner)
}

// Chat performs a synchronous non-streaming request.
func (c *LLMClient) Chat(messages []ollamaMessage) (string, error) {
	reqBody := ollamaRequest{
		Model:    c.model,
		Messages: messages,
		Stream:   false,
	}

	b, err := json.Marshal(reqBody)
	if err != nil {
		return "", err
	}

	resp, err := c.httpClient.Post(c.endpoint+"/api/chat", "application/json", bytes.NewReader(b))
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	var chatResp struct {
		Message struct {
			Content string `json:"content"`
		} `json:"message"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&chatResp); err != nil {
		return "", err
	}
	return chatResp.Message.Content, nil
}

// ModelName returns the active model name.
func (c *LLMClient) ModelName() string { return c.model }

File: UNIT-01/orchestrator/src/llm_client.rs
use crate::schema::directive_schema;
use crate::stream_parser::parse_stream;
use crate::types::{Directive, OllamaMessage, OllamaRequest};
use serde::Deserialize;
use std::time::Duration;

const OLLAMA_ENDPOINT: &str = "http://127.0.0.1:11434";

pub struct LLMClient {
    pub endpoint: String,
    pub model: String,
    client: reqwest::Client,
    show_cache: Option<ModelShowResponse>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct ModelShowResponse {
    pub modelfile: Option<String>,
    pub parameters: Option<String>,
    pub template: Option<String>,
    pub details: Option<ModelDetails>,
    pub model_info: Option<serde_json::Value>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct ModelDetails {
    pub parent_model: Option<String>,
    pub format: Option<String>,
    pub family: Option<String>,
    pub families: Option<Vec<String>>,
    pub parameter_size: Option<String>,
    pub quantization_level: Option<String>,
}

impl LLMClient {
    pub fn new(model: &str) -> Self {
        Self {
            endpoint: OLLAMA_ENDPOINT.to_string(),
            model: model.to_string(),
            client: reqwest::Client::builder()
                .timeout(Duration::from_secs(300))
                .build()
                .expect("reqwest client"),
            show_cache: None,
        }
    }

    pub async fn show_model(&mut self) -> Result<ModelShowResponse, String> {
        if let Some(ref cached) = self.show_cache {
            return Ok(cached.clone());
        }
        let body = serde_json::json!({"name": self.model});
        let resp = self
            .client
            .post(format!("{}/api/show", self.endpoint))
            .json(&body)
            .send()
            .await
            .map_err(|e| format!("http: {}", e))?;

        let show: ModelShowResponse = resp.json().await.map_err(|e| format!("json: {}", e))?;
        self.show_cache = Some(show.clone());
        Ok(show)
    }

    pub async fn stream_directives(
        &self,
        messages: Vec<OllamaMessage>,
        temperature: f64,
    ) -> Result<(Vec<Directive>, i32, i32), String> {
        let req = OllamaRequest {
            model: self.model.clone(),
            messages,
            stream: true,
            format: Some(directive_schema()),
            options: Some(
                serde_json::json!({"temperature": temperature})
                    .as_object()
                    .unwrap()
                    .clone(),
            ),
        };

        let resp = self
            .client
            .post(format!("{}/api/chat", self.endpoint))
            .json(&req)
            .send()
            .await
            .map_err(|e| format!("http: {}", e))?;

        let result = parse_stream(resp, None).await?;
        Ok((
            result.directives,
            result.prompt_tokens,
            result.output_tokens,
        ))
    }

    pub async fn stream_cli(
        &self,
        messages: Vec<OllamaMessage>,
        token_cb: &mut dyn FnMut(String),
        temperature: f64,
    ) -> Result<(String, Vec<Directive>, i32, i32), String> {
        let req = OllamaRequest {
            model: self.model.clone(),
            messages,
            stream: true,
            format: None,
            options: Some(
                serde_json::json!({"temperature": temperature})
                    .as_object()
                    .unwrap()
                    .clone(),
            ),
        };

        let resp = self
            .client
            .post(format!("{}/api/chat", self.endpoint))
            .json(&req)
            .send()
            .await
            .map_err(|e| format!("http: {}", e))?;

        let mut result = parse_stream(resp, Some(token_cb)).await?;
        if result.directives.is_empty() {
            result.directives = crate::stream_parser::extract_directives(&result.full_text);
        }
        Ok((
            result.full_text,
            result.directives,
            result.prompt_tokens,
            result.output_tokens,
        ))
    }

    pub async fn chat(&self, messages: Vec<OllamaMessage>) -> Result<String, String> {
        let req = OllamaRequest {
            model: self.model.clone(),
            messages,
            stream: false,
            format: None,
            options: None,
        };

        let resp = self
            .client
            .post(format!("{}/api/chat", self.endpoint))
            .json(&req)
            .send()
            .await
            .map_err(|e| format!("http: {}", e))?;

        #[derive(Deserialize)]
        struct ChatResp {
            message: ChatRespMessage,
        }
        #[derive(Deserialize)]
        struct ChatRespMessage {
            content: String,
        }

        let cr: ChatResp = resp.json().await.map_err(|e| format!("json: {}", e))?;
        Ok(cr.message.content)
    }
}
